language/stdlib/modules/libra_system.mvir

```diff
@@ -68,31 +86,59 @@ module LibraSystem {
 
     // Set the metadata for the current block.
     // The runtime always runs this before executing the transactions in a block.
-    // TODO: make this private, support other metadata
-    public block_prologue(height: u64) acquires BlockMetadata, ValidatorSet {
-      let height_ref: &mut u64;
-      let old_height: u64;
-
-      height_ref = &mut borrow_global_mut<BlockMetadata>(0xA550C18).height;
-      old_height = *copy(height_ref);
-      // ensure that height increases by 1
-      if (move(height) == copy(old_height) + 1) {
-          *move(height_ref) = move(old_height) + 1;
-          // triggers a reconfiguration if the validator keys or validator set has changed
-          Self.reconfigure();
-
-          return;
-      } else {
-          // TODO: standardize this error code
-          abort(99);
-      }
+    // TODO: 1. Make this private, support other metadata
+    //       2. Should the previous block votes be provided from BlockMetadata or should it come from the ValidatorSet
+    //          Resource?
+    public block_prologue(
+        timestamp: u64,
+        new_block_hash: bytearray,
+        previous_block_votes: bytearray,
+        proposer: address
+    ) acquires BlockMetadata, ValidatorSet{
+      Self.process_block_prologue(move(timestamp), move(new_block_hash), move(previous_block_votes), move(proposer));
+      // triggers a reconfiguration if the validator keys or validator set has changed
+      Self.reconfigure();
+      return;
+    }
+
+    // Update the BlockMetadata resource with the new blockmetada coming from the consensus.
+    process_block_prologue(
+        timestamp: u64,
+        new_block_hash: bytearray,
+        previous_block_votes: bytearray,
+        proposer: address
+    ) acquires BlockMetadata {
+        let block_metadata_ref: &mut Self.BlockMetadata;
+
+        block_metadata_ref =  borrow_global_mut<BlockMetadata>(0xA550C18);
+        *(&mut copy(block_metadata_ref).id) = move(new_block_hash);
+        // TODO: Should the prologue check if the timestamp is monotonically increased?
+        *(&mut copy(block_metadata_ref).timestamp) = move(timestamp);
+        *(&mut copy(block_metadata_ref).proposer) = move(proposer);
+        *(&mut copy(block_metadata_ref).height) = *(&copy(block_metadata_ref).height) + 1;
```

Comment:
```suggestion
        *(&mut move(block_metadata_ref).height) = *(&copy(block_metadata_ref).height) + 1;
```

Reply:
We can't do that as there's a block_metadata_ref on the RHS as well. We might try to optimize that pattern in the future at compiler level tho.